File: web/src/lib/branding.ts
// Author: Labiyb M. Said — DevSecOps Engineer
// Contact: saidlabiybm@gmail.com

export interface Brand {
  app_name: string;
  company: string;
  primary_hue: number;    // 0-360 — main CTAs, buttons, links, focus ring
  secondary_hue: number;  // 0-360 — badges, tags, secondary highlights
  surface_hue: number;    // 0-360 — subtle tint for panel/card/border surfaces
  logo_url: string;
  // "local" → email/password form  |  "oidc" → SSO redirect button
  auth_mode: "local" | "oidc";
}

export const DEFAULT_BRAND: Brand = {
  app_name: "DevPortal",
  company: "",
  primary_hue: 199,    // sky blue
  secondary_hue: 262,  // violet
  surface_hue: 215,    // cool gray-blue
  logo_url: "",
  auth_mode: "local",
};

// fetchBrand loads branding config from the Go server.
// Falls back to DEFAULT_BRAND on any network or parse error so the app
// always boots even if the endpoint is temporarily unavailable.
export async function fetchBrand(): Promise<Brand> {
  try {
    const res = await fetch("/branding.json");
    if (!res.ok) return DEFAULT_BRAND;
    const data = await res.json() as Brand;
    return { ...DEFAULT_BRAND, ...data };
  } catch {
    return DEFAULT_BRAND;
  }
}

// applyBrandCSS writes all three brand hues as CSS custom properties.
// Called once before React renders so there is no colour flash on load.
// index.css derives every colour token from these three properties via
// hsl(var(--brand-hue) ...) — the full palette shifts from a single call.
export function applyBrandCSS(brand: Brand): void {
  const root = document.documentElement;
  const clamp = (h: number) => Math.max(0, Math.min(360, Math.round(h)));
  root.style.setProperty("--brand-hue",           String(clamp(brand.primary_hue)));
  root.style.setProperty("--brand-secondary-hue", String(clamp(brand.secondary_hue)));
  root.style.setProperty("--brand-surface-hue",   String(clamp(brand.surface_hue)));
}

File: internal/handler/branding.go
// ---------------------------------------------------------------------------
// Author: Labiyb M. Said — DevSecOps Engineer
// Contact: saidlabiybm@gmail.com
// ---------------------------------------------------------------------------

// branding.go exposes GET /branding.json — a public endpoint that returns
// the portal's brand config read from environment variables.
//
// The React SPA fetches this once at boot (before first render) and applies
// the values as CSS custom properties and React context. The same Docker image
// ships to every customer; branding is purely env-var config.
//
// Response shape:
//
//	{
//	  "app_name":       "DevPortal",
//	  "company":        "Acme Corp",
//	  "primary_hue":    199,
//	  "secondary_hue":  262,
//	  "surface_hue":    215,
//	  "logo_url":       "https://cdn.example.com/logo.svg",
//	  "auth_mode":      "oidc"
//	}
//
// React reads all three hues and writes them as CSS custom properties before
// the first render — --brand-hue, --brand-secondary-hue, --brand-surface-hue.
// Every color token in index.css is derived from these three properties via
// hsl(var(--brand-*) ...) so the full palette shifts with a single env-var change.

package handler

import (
	"encoding/json"
	"net/http"
)

type brandingResponse struct {
	AppName      string `json:"app_name"`
	Company      string `json:"company"`
	PrimaryHue   int    `json:"primary_hue"`
	SecondaryHue int    `json:"secondary_hue"`
	SurfaceHue   int    `json:"surface_hue"`
	LogoURL      string `json:"logo_url"`
	// AuthMode tells the React SPA which login UI to render:
	// "local" → email/password form  |  "oidc" → SSO redirect button
	AuthMode string `json:"auth_mode"`
}

func (h *Handler) handleBranding(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	// 5-minute public cache — safe because branding only changes on redeploy.
	w.Header().Set("Cache-Control", "public, max-age=300")
	_ = json.NewEncoder(w).Encode(brandingResponse{
		AppName:      h.cfg.BrandAppName,
		Company:      h.cfg.BrandCompany,
		PrimaryHue:   h.cfg.BrandPrimaryHue,
		SecondaryHue: h.cfg.BrandSecondaryHue,
		SurfaceHue:   h.cfg.BrandSurfaceHue,
		LogoURL:      h.cfg.BrandLogoURL,
		AuthMode:     h.cfg.AuthMode,
	})
}
